cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -0,0 +1,239 @@
+import re
+from itertools import chain
+from pathlib import Path
+from typing import Iterable, Optional, Union
+from urllib.parse import urlparse
+
+from pyarn.lockfile import Lockfile
+from pyarn.lockfile import Package as PYarnPackage
+from pydantic import BaseModel
+
+from cachi2.core.errors import PackageRejected, UnexpectedFormat
+from cachi2.core.package_managers.npm import NPM_REGISTRY_CNAMES
+from cachi2.core.package_managers.yarn.project import PackageJson
+from cachi2.core.package_managers.yarn_classic.workspaces import (
+    Workspace,
+    extract_workspace_metadata,
+)
+from cachi2.core.rooted_path import RootedPath
+
+# https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/resolvers/exotics/git-resolver.js#L15-L17
+GIT_HOSTS = frozenset(("github.com", "gitlab.com", "bitbucket.com", "bitbucket.org"))
+GIT_PATTERN_MATCHERS = (
+    re.compile(r"^git:"),
+    re.compile(r"^git\+.+:"),
+    re.compile(r"^ssh:"),
+    re.compile(r"^https?:.+\.git$"),
+    re.compile(r"^https?:.+\.git#.+"),
+)
+
+
+class _BasePackage(BaseModel):
+    """A base Yarn 1.x package."""
+
+    name: str
+    version: Optional[str] = None
+    integrity: Optional[str] = None
+    dev: bool = False
+
+
+class _UrlMixin(BaseModel):
+    url: str
+
+
+class _RelpathMixin(BaseModel):
+    relpath: Path
+
+
+class RegistryPackage(_BasePackage):
+    """A Yarn 1.x package from the registry."""
```

Comment:
While working on generating SBOM I saw that there is an optional qualifier for `repository_url`:
https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs

The [default](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm) repository/registry for NPM is `https://registry.npmjs.org` so the second npm registry that we support is `"https://registry.yarnpkg.com` could/should be potentially included in the PURL.

Am I missing something ?

Reply:
Missed this, but you're right. We might want the resolved URL for registry packages when we get to the PURL generation. Added.